idea/src/volVis/vol-tfPrep/vol-tfPrep.1
.\" This is a comment
.\" The extra parameters on .TH show up in the headers
.TH vol-tfPrep 1 " " "NIST/ACMD/HPCVG" "HEV"
.SH NAME
vol-tfPrep - Prepare transfer function for direct volume rendering.

.SH SYNOPSIS
.B vol-tfPrep 
transferFuncFn  [ outputFnPrefix ]

.SH DESCRIPTION

\fBvol-tfPrep\fR 
reads a transfer function file and generates the GLSL source code
file(s) and the texture file(s) required for direct volume rendering
using the transfer function.

This program can be called independently, but it is also called
from the program vol-visBuilder.


.TP
.B transferFuncFn
The file that contains the transfer function definition.
The format of this file is described in the man
page \fIvol-transferFuncFmt\fR.

.TP
.B [ outputFnPrefix ]
If specified, this gives a prefix string from which the file names of the
generated GLSL and texture files are constructed.  
If this string ends with "/", then the files are placed in the indicated
directory.  If the string does not end in "/", then the output file
names are generated by appending appropriate strings to the end of
this string.  
If this is not specified, then the output files are placed in the
current directory.

.SH OPTIONS
.TP
.B --volume  volumeFileName
Specifies the name of a volume data set in \fI.mha\fR format that
may be needed in the preparation of the the transfer function.
Currently only the \fIGkBoundary\fR transfer functions can use this
option; see the man page for \fIvol-transferFuncFmt\fR.  

.SH OUTPUT

This program generates two types of files:

.RS
* GLSL source code:  The file name will end in \fI.glsl\fR. This code should typically be included in the fragment shader.

* Texture files: The texture files will be in \fI.mha\fR format.  Textures are used by the GLSL code where they appear as \fIsampler\fR variables.
.RE

In addition, this program writes information to standard output about the
files produced and some resources required by the GLSL code.
This information is formatted for parsing by other software 
(such as vol-visBuilder), but it is 
(we hope) human readable.  This output is formatted into a series of
lines; each line is in one of the following forms:

.TP
.I GLSL  fileName.glsl  functionName
This indicates that GLSL source code containing a function 
named \fIfunctionName\fR has been written to a file named \fIfileName.glsl\fR.

.TP
.I TEXTURE txtrFileName.mha glslVarName nChan nDim [ min max  min max...]
This indicates that a texture has been written to the file 
\fItxtrFileName.mha\fR and
this texture corresponds to the GLSL variable \fIglslVarName\fR in the
generated GLSL source code. The texture has \fInChan\fR channels 
(components) per element and has \fInDim\fR dimensions.  
The data values used to do the texture lookup are expected 
to fall into the ranges specified by the \fImin/max\fR pairs.  
There should be one min/max pair for each dimension.

.TP
.I REQ_EXTERNAL stdName  objType  dataType  glslName
This indicates that the GLSL code for the transfer function requires
an external variable or function (beyond the GLSL built-ins).
.RS
.br
\fIstdName\fR gives a standardized name for the external
.br
\fIobjType\fR indicates whether it is a function or variable
.br
\fIdataType\fR gives the GLSL data type of the external
.br
\fIglslName\fR gives its name in the generated GLSL source code.
.RE






.SH EXAMPLES

.B vol-tfPrep $HEVROOT/idea/src/volVis/testData/atmos.tf myVis

This command prepares the transfer function described in the file 
\fIatmos.tf\fR by generating the GLSL source file \fImyVis.CO2D.glsl\fR
and the texture file \fImyVis.CO2D0000.mha\fR.

This command generates the following output (in the format described above):

.nf
    GLSL myVis.CO2D.glsl CO2D
    TEXTURE myVis.CO2D0000.mha CO2D_RgbaLut1D 4 1  -20 1
    REQ_EXTERNAL  volGradCalc  func  vec3   volGradCalc
    REQ_EXTERNAL  volPos       func  vec3   volPos
    REQ_EXTERNAL  volPosPrev   func  vec3   volPosPrev
    REQ_EXTERNAL  volScalar    func  float  volScalar
.fi

Note that the GLSL code in \fImyVis.CO2D.glsl\fR makes use of the
externals specified by the \fIREQ_EXTERNAL\fR lines.



.SH SEE ALSO
vol-visBuilder, hev-convertDataArray, vol-transferFuncFmt




idea/src/volVis/vol-tfPrep/vol-tfPrep.cpp
//
// vol-tfPrep  transferFuncFN  outputFNPrefix
//
// outputFNPrefix: if it ends in / it's assumed to be a directory name
// and the output file names will be placed in that directory.

#include <stdio.h>
#include <stdlib.h>

#include <dtk.h>
#include <VolVis.h>

static void
usage ()
        {
        fprintf (stderr, 
        "Usage:  vol-tfPrep  [--volume volFN] transferFuncFN  [outFnPrefix]\n");
        }
        

static int
parseArgs (int argc, char **argv, 
            std::string & volumeName, 
            std::string & transferFuncFn, 
            std::string & outFnPrefix )
    {
    volumeName = "";
    transferFuncFn = "";
    outFnPrefix = "";

    if ( (argc < 2) || (argc > 5) )
        {
        dtkMsg.add (DTKMSG_ERROR,
                "vol-tfPrep: Bad argument count.\n");
        usage ();
        return -1;
        }

    bool badArgs = false;

    switch (argc)
        {
        case 2:
            transferFuncFn = argv[1];
            break;

        case 3:
            transferFuncFn = argv[1];
            outFnPrefix = argv[2];
            break;

        case 4:
            if ( (strncasecmp ("--volume", argv[1], strlen(argv[1])) == 0) && 
                 (strlen(argv[1]) > 3) )
                {
                volumeName = argv[2];
                transferFuncFn = argv[3];
                }
            else
                {
                badArgs = true;
                }
            break;

        case 5:
            if ( (strncasecmp ("--volume", argv[1], strlen(argv[1])) == 0) && 
                 (strlen(argv[1]) > 3) )
                {
                volumeName = argv[2];
                transferFuncFn = argv[3];
                outFnPrefix = argv[4];
                }
            else
                {
                badArgs = true;
                }
            break;

        }


        if (badArgs)
            {
            dtkMsg.add (DTKMSG_ERROR,
                "vol-tfPrep: Bad arguments.\n");
            usage ();
            return -1;
            }


    if (outFnPrefix != "")
        {
        int len = outFnPrefix.size();
        if ( (outFnPrefix[len-1] != '/') && (outFnPrefix[len-1] != '.') )
            {
            // the idea is that the prefix is separated by what come after with a .
            outFnPrefix += ".";
            }
        }

    return 0;
    }  // end of parseArgs

int
main (int argc, char **argv)
        {


        // send messages to stderr
        dtkMsg.setFile(stderr) ;


#if 0
        if ( (argc < 2) || (argc > 3) )
                {
                dtkMsg.add (DTKMSG_ERROR,
                        "vol-tfPrep: Bad argument count.\n");
                usage ();
                return -1;
                }

        std::string transferFuncFn = argv[1];

        std::string outFnPrefix;
        if (argc == 3)
                {
                outFnPrefix = argv[2];
                int len = outFnPrefix.size();
                if (outFnPrefix[len-1] != '/')
                        {
                        outFnPrefix += ".";
                        }
                }
        else
                {
                outFnPrefix = "";
                }
#else
        std::string volumeFn;
        std::string transferFuncFn;
        std::string outFnPrefix;

        if (parseArgs (argc, argv, volumeFn, transferFuncFn, outFnPrefix ))
            {
            return -1;
            }

#endif
                


        VolVis::TransferFunction * tf = VolVis::readTF (transferFuncFn);

        if (tf == NULL)
                {
                dtkMsg.add (DTKMSG_ERROR,
                    "vol-tfPrep: Error reading transfer function file %s.\n", 
                    transferFuncFn.c_str());
                    return -1;
                }

        if ((volumeFn != "") && (volumeFn != "-none"))
            {
            tf->setInputDataFileName (volumeFn);
            }

        std::string funcName = tf->getFuncName();


        std::string glslName = outFnPrefix + funcName + ".glsl";
        if (tf->writeGLSL (glslName))
                {
                dtkMsg.add (DTKMSG_ERROR,
                        "vol-tfPrep: Error writing GLSL file %s.\n", 
                        glslName.c_str());
                return -1;
                }

        printf ("GLSL %s %s\n", glslName.c_str(), funcName.c_str());

        int numTextures = tf->getNumTextures ();
        if (numTextures < 0)
            {
            dtkMsg.add (DTKMSG_ERROR,
             "vol-tfPrep: Error creating textures for transfer function  %s.\n",
             transferFuncFn.c_str());
            return -1;
            }

        for (int i = 0; i < numTextures; i++)
                {
                char seqstr[10];
                sprintf (seqstr, "%04d", i);
                std::string glslFn = outFnPrefix + funcName + seqstr + ".mha";
                if (tf->writeTexture (i, glslFn))
                        {
                        dtkMsg.add (DTKMSG_ERROR,
                                "vol-tfPrep: Error writing texture file %s.\n", 
                                glslFn.c_str());
                        return -1;
                        }
                std::string glslVarName;
                int nChan;
                int nDim;
                float range[2];

                if (tf->getTextureInfo (i, glslVarName, nChan, nDim))
                        {
                        dtkMsg.add (DTKMSG_ERROR,
                            "vol-tfPrep: Error getting texture info "
                            "for transfer function file %s.\n", 
                                glslFn.c_str());
                        return -1;
                        }


                printf ("TEXTURE %s %s %d %d ", 
                        glslFn.c_str(), glslVarName.c_str(), nChan, nDim);

                
                float *ranges = new float [nDim*2];
                if (tf->getTextureRanges (i, ranges))
                    {
                    dtkMsg.add (DTKMSG_ERROR,
                            "vol-tfPrep: Error getting texture range info "
                            "for transfer function file %s.\n", 
                                glslFn.c_str());
                    return -1;
                    }
                for (int j = 0; j < nDim; j++)
                    {
                    printf ("   %.9g %.9g ", ranges[j*2+0], ranges[j*2+1]);
                    }
                printf ("\n");
                delete ranges;

                }  // end of loop over textures



        // Now print out a list of the required externals
        const std::map <std::string, VolVis::TFReqExt *> & reqs
                        = tf->getRequiredExts ();

        std::map<std::string, VolVis::TFReqExt *>::const_iterator re;
        for (re = reqs.begin(); re != reqs.end(); re++)
                {
                printf ("REQ_EXTERNAL   %-12s  %10s   %-8s  %s",
                    re->second->stdName.c_str(), re->second->objType.c_str(),
                    re->second->dataType.c_str(), re->second->glslName.c_str());
                for (int j = 0; j < re->second->infoStrs.size(); j++)
                    {
                    printf (" %s", re->second->infoStrs[j].c_str());
                    }
                printf ("\n");
                }



        return 0;
        }  // end of main






//
// vol-tfPrep  transferFuncFN  outputFNPrefix
//
// outputFNPrefix: if it ends in / it's assumed to be a directory name
// and the output file names will be placed in that directory.


